### ISSUE TITLE
pg_advisory_lock fails for knex

### ISSUE THREAD

**Issue by wingedfox** [tgriesser/knex#1566]
Hi,

I'm trying to use [pg_advisory_lock](https://www.postgresql.org/docs/9.1/static/explicit-locking.html), but in order to make it working properly I have to setup new connection for every bunch of locked queries.

```js
const startTime = Date.now()
const id = Math.random()

const _knex = new knex(config.get('db_server'))   // this works
//const _knex = require('./knexInstance')           // this utilizes single session and reuse locks

console.log(id, 1, Date.now() - startTime)
        Promise.all([
            _knex.raw('select pg_try_advisory_lock(1,1);')
                .then((data) => {
                    console.log(id, 2, Date.now() - startTime, 'one lock!: ', data.rows)
                    return true
                })
                .then(() => Promise.all([
                    getTimeout('All', 10000)
                        .then((message) => {
                            console.log(id, 3, Date.now() - startTime, message)
                        })
                ]))
                .then(() => _knex.raw('select pg_advisory_unlock(1,1);'))
        ])
            .then(() => {
                console.log(id, 4, Date.now() - startTime)

                response.json({})
            })

        function getTimeout (message, time) {
            return new Promise((resolve, reject) => {
                setTimeout(resolve, time, 'Block! ' + message)
            })
        }
```

Is there a good way to setup different sessions for different queries in knex to reuse pooling?

I know that I can use transactions, but I'd prefer to avoid using them.

**Comment by elhigu** [comment]
You can make multiple knex instances without connection pooling or pool size 1 for those queries which should be sent to specific connection. I've never tried that though.

**Comment by wingedfox** [comment]
Thanks for suggestion, this is exactly the thing I have to do now. But it is kind of a performance killer, setting up a TCP connection is one of the most expensive tasks.

**Comment by elhigu** [comment]
I believe that poolsize 1 actually still reuses that one connection between requests, so you don't have to make new tcp connections all the time... But with that you would kind of need to create pool of knex instances with single connection, which sounds pretty annoying.

Anyways if you like to have this feature, I could set this issue as feature request. Probably similar API to transaction API could work for allowing to sending multiple queries to same connection.

**Comment by wingedfox** [comment]
You're right.
I need to have a new API to manage database sessions while reusing connections.

It would be definitely great to be able to bind single session to particular business logic like it is done with the transactions.

**Comment by wingedfox** [comment]
@elhigu 
This issue should not be closed this way. 

1. The transactions are NOT the locks.
2. There are pg_advisory_lock and pg_advisory_xact_lock
3. There's the official clause that you should not use plain locks with transactions

Please study how the locks work: https://vladmihalcea.com/how-do-postgresql-advisory-locks-work/

This issue should be solved with either proper feature request or with a statement that knex has no plans to support the locks.

**Comment by wingedfox** [comment]
@elhigu there's no need to implement this for a single dialect. The problem is KNEX takes new connection from pool for each query which makes advisory lock useless.

As long as locks are bound to sessions there should be a way to 'lock' a connection in the same manner as transaction does. But there should be the explicit API for that.

And yes, I don't like the idea of using transactions in place of the lightweight locking.

**Comment by elhigu** [comment]
@omeid yes, in transaction you should use `pg_advisory_xact_lock` my point being, if you like to use locks like transactions, why not use real transactions + xact locks instead. 

Maybe there is a valid use case for needing that, but I'm failing to see it (usually its just because of bad design and causes unexpect problems, if people are not super careful with their code). If its just micro-optimization to prevent need for extra `begin / commit` queries, it might be significant in some rare cases...

Anyways I repeat and emphasize from my last comment:

> If there is need for a feature to have an additional API for advisory locks (which should be as much DB agnostic as possible) then **proper feature request is needed**.

and for @wingedfox 

> Current way to use them is to use knex.raw() and the other issue of **not being able to select connection where to send queries is a separate issue**.
